constantine/math/arithmetic/finite_fields_square_root_precomp.nim

```diff
@@ -0,0 +1,193 @@
+
+  
+# sqrtAlg_GetPrecomputedRootOfUnity sets target to g^(multiplier << (order * sqrtParam_BlockSize)), where g is the fixed primitive 2^32th root of unity.
+#
+# We assume that order 0 <= order*sqrtParam_BlockSize <= 32 and that multiplier is in [0, 1 <<sqrtParam_BlockSize)
+func sqrtAlg_GetPrecomputedRootOfUnity(target: var Fp, multiplier: int, order: uint) =
+  target = Fp.C.sqrtDlog(PrecomputedBlocks)[order][multiplier]
+
+
+func sqrtAlg_ComputeRelevantPowers(z: Fp, squareRootCandidate: var Fp, rootOfUnity: var Fp) {.addchain.} =
+  ## sliding window-type algorithm with window-size 5
+  ## Note that we precompute and use z^255 multiple times (even though it's not size 5)
+  ## and some windows actually overlap
+  var z2, z3, z7, z6, z9, z11, z13, z19, z21, z25, z27, z29, z31, z255 {.noInit.} : Fp
+  var acc: Fp
+  z2.square(z)
+  z3.prod(z2, z)
+  z6.prod(z3, z3)
+  z7.prod(z6, z)
+  z9.prod(z7, z2)
+  z11.prod(z9, z2)
+  z13.prod(z11, z2)
+  z19.prod(z13, z6)
+  z21.prod(z19, z2)
+  z25.prod(z19, z6)
+  z27.prod(z25, z2)
+  z29.prod(z27, z2)
+  z31.prod(z29, z2)
+  acc.prod(z27, z29)
+  acc.prod(acc, acc)
+  acc.prod(acc, acc)
+  z255.prod(acc, z31)
+  acc.prod(acc, acc)
+  acc.prod(acc, acc)
+  acc.prod(acc, z31)
+  acc.square_repeated(6)
+  acc.prod(acc, z27)
+  acc.square_repeated(6)
+  acc.prod(acc, z19)
+  acc.square_repeated(5)
+  acc.prod(acc, z21)
+  acc.square_repeated(7)
+  acc.prod(acc, z25)
+  acc.square_repeated(6)
+  acc.prod(acc, z19)
+  acc.square_repeated(5)
+  acc.prod(acc, z7)
+  acc.square_repeated(5)
+  acc.prod(acc, z11)
+  acc.square_repeated(5)
+  acc.prod(acc, z29)
+  acc.square_repeated(5)
+  acc.prod(acc, z9)
+  acc.square_repeated(7)
+  acc.prod(acc, z3)
+  acc.square_repeated(7)
+  acc.prod(acc, z25)
+  acc.square_repeated(5)
+  acc.prod(acc, z25)
+  acc.square_repeated(5)
+  acc.prod(acc, z27)
+  acc.square_repeated(8)
+  acc.prod(acc, z)
+  acc.square_repeated(8)
+  acc.prod(acc, z)
+  acc.square_repeated(6)
+  acc.prod(acc, z13)
+  acc.square_repeated(7)
+  acc.prod(acc, z7)
+  acc.square_repeated(3)
+  acc.prod(acc, z3)
+  acc.square_repeated(13)
+  acc.prod(acc, z21)
+  acc.square_repeated(5)
+  acc.prod(acc, z9)
+  acc.square_repeated(5)
+  acc.prod(acc, z27)
+  acc.square_repeated(5)
+  acc.prod(acc, z27)
+  acc.square_repeated(5)
+  acc.prod(acc, z9)
+  acc.square_repeated(10)
+  acc.prod(acc, z)
+  acc.square_repeated(7)
+  acc.prod(acc, z255)
+  acc.square_repeated(8)
+  acc.prod(acc, z255)
+  acc.square_repeated(6)
+  acc.prod(acc, z11)
+  acc.square_repeated(9)
+  acc.prod(acc, z255)
+  acc.square_repeated(2)
+  acc.prod(acc, z)
+  acc.square_repeated(7)
+  acc.prod(acc, z255)
+  acc.square_repeated(8)
+  acc.prod(acc, z255)
+  acc.square_repeated(8)
+  acc.prod(acc, z255)
+  acc.square_repeated(8)
+  acc.prod(acc, z255)
+  # acc is now z^((BaseFieldMultiplicativeOddOrder - 1)/2)
+  rootOfUnity.square(acc)
+  rootOfUnity *= z
+  squareRootCandidate.prod(acc, z)
```

Comment:
For future reference, I have a feeling that this addchain computes the same thing as https://github.com/mratsim/constantine/blob/973e0aaf9b3a7ce185d30750afc8ee12dc042cd6/constantine/math/constants/bandersnatch_sqrt.nim#L19-L149

Also it would need to be moved to Banderwagon/Bandersnatch specific files in the future